### ISSUE TITLE
Custom resource configuration should support references to Secrets

### ISSUE THREAD

**Issue by rafal-roch** [Kong/kubernetes-ingress-controller#299]
NOTE: GitHub issues are reserved for **bug reports only**.
For anything else, please join the conversation
in Kong Nation https://discuss.konghq.com/c/kubernetes.

---

### Summary

KongCredential object as well as some plugins require passing sensitive information in their configuration, i.e. passwords, access tokens, etc.

```
apiVersion: configuration.konghq.com/v1
kind: KongCredential
metadata:
  name: credential-team-x
consumerRef: consumer-team-x
type: key-auth
config:
  key: 62eb165c070a41d5c1b58d9d3d725ca1
```

It would be great if it was possible to reference Kubernetes Secrets here, instead of passing value as plaintext, e.g.:

```
apiVersion: configuration.konghq.com/v1
kind: KongCredential
metadata:
  name: credential-team-x
consumerRef: consumer-team-x
type: key-auth
config:
  key: 
     valueFrom:
        secretKeyRef:
          name: my-secret
          key: secret_key
```
**Kong Ingress controller version**
0.3.0

**Kong or Kong Enterprise version**
1.1.1

### Kubernetes version

**paste `kubectl version` output**

### Environment

- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

### Steps To Reproduce

1.
2.
3.
4.

**Comment by hbagdi** [comment]
@rafal-roch This is something that comes up often and is certainly on the roadmap. We will track the progress of this feature request using this issue. Thanks!

**Comment by ismailbaskin** [comment]
What do you think about plain `Secret` with kong annotations or labels?

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: credential-team-x
  annotations:
    type.konghq.com: key-auth
    consumerRef.konghq.com: consumer-team-x
data:
  key: NjJlYjE2NWMwNzBhNDFkNWMxYjU4ZDlkM2Q3MjVjYTE=
```

**Comment by timduhenchanter** [comment]
Also please do not limit the scope to just KongCredential CRD.  Valid use cases with other CRD types.  CRD config values from secret should be generalized across all supported CRDs.

**Comment by hbagdi** [comment]
@ismailbaskin I've been thinking about this for a while now and my thought is close to what you are thinking but I'm thinking of turning the relationship the other way around, where the consumer refers to the secret, rather than the secret referring to the consumer:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: team-x-apikey
data:
  key: NjJlYjE2NWMwNzBhNDFkNWMxYjU4ZDlkM2Q3MjVjYTE=
```

And then the consumer:

```yaml
apiVersion: configuration.konghq.com/v1
kind: KongConsumer
metadata:
  name: harry
  annotations:
     credentials.konghq.com: team-x-apikey
username: harry
```

It might make sense to not add an annotation but rather `credentials` as an array into the KongConsumer CRD itself.
Thoughts?

**Comment by ismailbaskin** [comment]
This relation idea is much better. I don't find it right to use annotation in CRDs in general. Annotations are hacky and useful only when you have no control over the resource definition (e.g. Ingress) 

And how do you intend to detect the kong credential type (`key-auth`)? 
My suggestion is 
```yaml
apiVersion: configuration.konghq.com/v2
kind: KongConsumer
metadata:
  name: harry
spec:
  username: harry
  credentials:
    - type: key-auth
      secretRef:
        name: team-x-apikey
```

**Comment by hbagdi** [comment]
>  What do you think about this example?

I like this.

The only thing is that the credential type should belong to the secret created and then those secrets are associated with KongConsumer as you mentioned above.
The type of credential can be part of the secret itself.

Thoughts?

**Comment by hbagdi** [comment]
Update: #430 fixes this issue.

**Comment by jaskiratr** [comment]
@hbagdi Thanks for having this implemented! Is this feature available in `kong-ingress-controller:0.6.2`?
Can you kindly share an example on what's the proper way for using secrets?

**Comment by hbagdi** [comment]
This will be shipped in 0.7.0 (not released yet).
